runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -298,15 +298,25 @@ function M.on_diagnostic(_, result, ctx, config)
   local uri = ctx.params.textDocument.uri
   local fname = vim.uri_to_fname(uri)
 
-  if result == nil then
-    return
-  end
-
-  if result.kind == 'unchanged' then
-    return
+  local diagnostics = {}
+  if result ~= nil then
```

Comment:
This PR does change the behaviour.

When result is nil, `diagnostics` (on line 301) will now be an empty table.

On line 320 we only exit early if there is no buffer attached for the file.

Otherwise we will continue down to the end of this function where on line 346 we call `vim.diagnostic.set` with an empty table, which clears the diagnostics.

I'll add a test for this.

Reply:
I see, thanks for detailed explanation. `#diagnostics == 0` **and** `vim.fn.bufexists(fname) == 0`.

Actually I am not convinced that the handler should not behave so, when LSP responds an invalid response `nil`. To me, it feels like an over-correction specific to the behavior of the ruby LSP. What if it were meant to be "unchanged"? As a workaround, provided the LSP server is yet to be fixed, user *may* override the vim.lsp.handler to change the semantics of `nil` as "empty diagnostics".

Reply:
That's an excellent point. It's definitely possible that another possible interpretation of `nil` would be "unchanged".

What would it look like for the user to override the handler for a specific lsp to change the semantics for `nil`?

Reply:
I've left comment for one example outside this thread.